### PR TITLE
Generate Code Cities using the Language Server Protocol

### PR DESCRIPTION
**Pull request by falko17**
# Summary
This PR implements the main part of my master's thesis: It allows users to generate code cities based on information provided by LSP language servers. (Integration of LSP beyond that, e.g., in code windows, is not yet implemented.)
There are also several other changes, such as a determinate progress spinner—see details below.

# Supported languages
Below are all the languages SEE supports now. For all of these languages, there's at least one language server configured in `LSPServer.cs` for which I've tested and confirmed that code cities can be generated with it.
For each language server, a link has been provided to installation instructions for it. It will be shown if it is detected that the server is not installed.

## Programming languages
* C
* C++
* C#
* Dart
* Go
* Haskell
* Java
* JavaScript
* Kotlin
* $\LaTeX$
* Lua
* MATLAB
* PHP
* Python
* Ruby
* Rust
* TypeScript
* Zig

## Miscellaneous languages
* JSON
* LaTeX
* Markdown
* XML

> [!NOTE]
> These are just the languages I've tested. If there are any other languages for which there's interest of adding support to SEE, it should be rather simple to add them to `LSPServer.cs`.

# Details
- A new `LSPGraphProvider` has been added, along with various related components (`LSPImporter`, `LSPHandler`, `LSPLanguage`, `LSPServer`). This makes up the majority of the changes for this pull request and is the biggest step of #686.
   - `LSPLanguage` and `LSPServer` are purely data models to represent languages and language servers for the user to choose from, respectively.
   - `LSPHandler` handles the actual communication with the LSP server and abstracts over/wraps OmniSharp's LSP client.
   - `LSPImporter` contains the actual main algorithm responsible for creating a code city, using a previously set up `LSPHandler`.
   - I won't go into more detail on each of these here, as it would take up a lot of space, and I plan to do it in the master's thesis anyway.
   - An important caveat: Due to a bug in OmniSharp that seems time-consuming to fix for now, we cannot retrieve the call hierarchy (i.e., outgoing calls) for nodes. Hence, these will only show up as "reference" edges.
- For the LSP import to work efficiently, an [interval tree](https://en.wikipedia.org/wiki/Interval_tree) has been implemented as an augmented [KD-tree](https://en.wikipedia.org/wiki/K-d_tree).
   - This is used to match node ranges to the most "fitting" nodes.
   - Tests have been added to make sure this works correctly.
- A progress bar has been added in the editor for code cities. This can be used to signal the progress of an ongoing graph creation process to the user.
   - For this purpose, a callback has been added that graph providers can use to report their progress.
- A determinate progress spinner (i.e., one that updates based on progress reports) has been implemented. It can be used as `LoadingSpinner.ShowDeterminate("Loading something...", out Action<float> updateProgress`, where `updateProgress` can be called with a float from 0 to 1 to indicate the progress percentage.
   - For loading processes in general, the process name is now shown below the loading spinner rather than when hovering above it. This makes it more legible to the user what is happening.
- Cooperative cancelling has been implemented for graph providers. That is, ongoing asynchronous processes can now be cancelled using a `CancellationToken`. Currently, loading processes can be canceled by clicking on "Reset data".
- A general representation of Ranges has been implemented into `Attributable`. Internally, these are represented as four integers (start line, end line, start character, end character).
- The containment logic for ranges has been fixed, as there were previously some subtle errors. Tests have been written to catch such errors in the future.
   - Additionally, ranges are now (non-transitively) comparable in order to find out which of two ranges takes up more characters. 
- A workaround for an upstream bug in the CI has been implemented by reverting to an earlier version of the `github-script` dependency.
- Unity has been updated to 2022.3.26f1.
- As a result of the above changes, the following NuGet dependencies have been added:
   - [OmniSharp's LSP implementation](omnisharp/csharp-language-server-protocol) to use as an LSP client
   - [Supercluster.KDTree](https://github.com/ericreg/Supercluster.KDTree) for an efficient kd-tree implementation for the LSP import algorithm. It has been augmented for use as an interval tree in the class `KDIntervalTree<E>`.
   - [Markdig](https://github.com/xoofx/markdig) as a Markdown processor for LSP-provided user-facing text. Currently, the markdown is simply converted to plaintext, but in the future, we can use Markdig to convert it to [TextMeshPro rich text tags](https://docs.unity3d.com/Packages/com.unity.textmeshpro@4.0/manual/RichTextSupportedTags.html).

### CODE REVIEW

File: Assets/SEE/DataModel/DG/IO/LSPImporter.cs [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+                    {
+                        await ConnectNodeViaAsync(Handler.Implementation, "Implementation_Of", node, graph, reverseDirection: true, token);
+                    }
+                    if (IncludeEdgeTypes.HasFlag(EdgeKind.Reference) && Handler.ServerCapabilities.ReferencesProvider.TrueOrValue())
+                    {
+                        await ConnectNodeViaAsync((p, line, character) => Handler.References(p, line, character), "Reference", node, graph, reverseDirection: true, token);
+                    }
+                    if (IncludeEdgeTypes.HasFlag(EdgeKind.Call) && Handler.ServerCapabilities.CallHierarchyProvider.TrueOrValue())
+                    {
+                        // FIXME (UniTask-internal problem, sends wrong method name):
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/DG/IO/LSPImporter.cs:273]: Done (marked as external FIXME).

File: Assets/SEE/DataModel/DG/IO/LSPImporter.cs [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+                        return $"```{x.Language}\n{x.Value}\n```";
+                    }
+                    else
+                    {
+                        return x.Value;
+                    }
+                }));
+            }
+
+            // TODO: Parse markdown to TextMeshPro rich text (custom MarkDig parser).
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/DataModel/DG/IO/LSPImporter.cs]: Done (#728).

File: Assets/SEE/GraphProviders/GraphProvider.cs [hunk 3]
```diff
@@ -24,8 +25,14 @@ public abstract class GraphProvider
...
         /// <param name="city">settings possibly necessary to provide a graph</param>
+        /// <param name="changePercentage">callback to report progress from 0 to 1</param>
+        /// <param name="token">cancellation token</param>
         /// <returns>provided graph based on <paramref name="graph"/></returns>
-        public abstract UniTask<Graph> ProvideAsync(Graph graph, AbstractSEECity city);
+        /// <exception cref="OperationCanceledException">thrown if the operation is cancelled</exception>
+        public abstract UniTask<Graph> ProvideAsync(Graph graph, AbstractSEECity city,
+                                                    Action<float> changePercentage = null,
+                                                    CancellationToken token = default);
+        // TODO: Implement changePercentage for providers other than LSP and Pipeline.
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/GraphProviders/GraphProvider.cs]: Done (#729).

File: Assets/SEE/Tools/LSP/LSPHandler.cs [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+                Implementation = new ImplementationCapability
+                {
+                    LinkSupport = false
+                },
+                References = new ReferenceCapability(),
+                CallHierarchy = new CallHierarchyCapability(),
+                TypeHierarchy = new TypeHierarchyCapability(),
+                Hover = new HoverCapability
+                {
+                    // TODO: Switch this once we have proper markdown support.
```
[the remark this answers is not in the document]
Comment (falko17) [reply Assets/SEE/Tools/LSP/LSPHandler.cs]: Done (#728).

File: Assets/SEE/DataModel/DG/IO/LSPImporter.cs [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+        // By default, all edge types are included, except for <see cref="EdgeKind.Definition"/>
+        // and <see cref="EdgeKind.Declaration"/>, since nodes would otherwise often get a self-reference.
+        EdgeKind IncludeEdgeTypes = EdgeKind.All & ~(EdgeKind.Definition | EdgeKind.Declaration),
+        bool AvoidSelfReferences = true,
+        bool AvoidParentReferences = true)
+    {
+        /// <summary>
+        /// A mapping from directory paths to their corresponding nodes.
+        /// </summary>
+        private readonly IDictionary<string, Node> NodeAtDirectory = new Dictionary<string, Node>();
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/DataModel/DG/IO/LSPImporter.cs]: I fixed it.

File: Assets/SEE/DataModel/DG/IO/LSPImporter.cs [hunk 6]
```diff
@@ -0,0 +1,716 @@
...
+
+        /// <summary>
+        /// A mapping from file paths to their corresponding range trees.
+        /// </summary>
+        /// <seealso cref="KDIntervalTree{T}"/>
+        private IDictionary<string, KDIntervalTree<Node>> RangeTrees = new Dictionary<string, KDIntervalTree<Node>>();
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEE/DataModel/DG/IO/LSPImporter.cs]: I fixed it.

File: Assets/SEE/Tools/LSP/LSPHandler.cs [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+            switch (messageParams.Type)
+            {
+                case MessageType.Error:
+                    Debug.LogError(messageParams.Message);
+                    break;
+                case MessageType.Warning:
+                    Debug.LogWarning(messageParams.Message);
+                    break;
+                case MessageType.Info:
+                case MessageType.Log:
```
Comment (koschke) [inline Assets/SEE/Tools/LSP/LSPHandler.cs]: You can use an `or` between two cases: `case MessageType.Info or MessageType.Log:`.
That is more readable.

Comment (koschke) [reply Assets/SEE/Tools/LSP/LSPHandler.cs]: I fixed it.

File: Assets/SEE/Utils/TeeStream.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    /// </summary>
+    /// <remarks>
+    /// The code for this class has been created with the help of GPT-4.
+    /// </remarks>
+    public class TeeStream : Stream
+    {
+        /// <summary>
+        /// The primary stream, which is the stream to read from.
+        /// </summary>
+        private readonly Stream _primaryStream;
```
Comment (koschke) [inline Assets/SEE/Utils/TeeStream.cs]: A name should not start with an underscore.

Comment (koschke) [reply Assets/SEE/Utils/TeeStream.cs]: I fixed it.

File: Assets/SEE/Utils/TeeStream.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+        /// <summary>
+        /// The secondary stream to write the <see cref="_primaryStream"/> to.
+        /// </summary>
+        private readonly Stream _secondaryStream;
```
Comment (koschke) [inline Assets/SEE/Utils/TeeStream.cs]: A name should not start with an underscore.

Comment (koschke) [reply Assets/SEE/Utils/TeeStream.cs]: I fixed it.

File: Assets/SEETests/RangeExamples.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
+using System.Collections.Generic;
+using SEE.DataModel.DG;
+
+namespace SEETests
```
Comment (koschke) [inline Assets/SEETests/RangeExamples.cs]: Should be in name space `SEE.DataModel.DG`.

Comment (koschke) [reply Assets/SEETests/RangeExamples.cs]: I fixed it.

**Review by koschke — CHANGES_REQUESTED** [review]
I had only a few minor suggestions, which I implemented myself.

**Review by koschke — APPROVED** [review]

**Pull request merged**